# Тести для малих вибірок та пропорцій



**Завдання 1**. E-commerce компанія після редизайну сайту підозрює, що **середній час до покупки (time-to-purchase)** користувачів **збільшився**.
Історично середній час від першого заходу на сайт до покупки становив $\mu_0 = 123.7$ хвилин.

Команда аналітиків випадково обрала дані **7 покупців після редизайну**:
`128, 135, 121, 142, 126, 151, 123`

З допомогою наявних даних зʼясуйте, чи збільшився середній час покупки після редизайну.

Для цього виконайте наступні 6 кроків. Правильне виконання кожного з кроків оцінюється в 1 бал.

1. Запишіть нульову та альтернативну гіпотези і визначте тип тесту.
2. Обчисліть вибіркові статистики: $\bar x$, $s$, $n$.
3. Оберіть тип тесту та виконайте його (знайдіть критичне значення тесту, статистику тесту та р-значення) будь-яким пасуючим способом, щоб перевірити гіпотезу на рівні значущості $\alpha = 0.10$.
4. Прийміть рішення, чи відхиляєте ви гіпотезу $H_0$ використовуючи p-value.
5. Напишіть висновок: чи справді редизайн сайту подовжив час до покупки?
6. Чи зміниться ваше рішення при зміні рівня значущості на $\alpha = 0.05$.


In [1]:
import numpy as np
from scipy import stats

time_to_purchase = np.array([128, 135, 121, 142, 126, 151, 123])
mu0 = 123.7
alpha = 0.10

Оскільки перевіряємо чи час збільшився:

$H_0: μ = 123.7$

$H_a: μ > 123.7$

Це правосторонній t-тест.

In [2]:
x_bar = np.mean(time_to_purchase)
s = np.std(time_to_purchase, ddof=1)
n = len(time_to_purchase)

print(f"Вибіркове середнє: {x_bar:.2f}")
print(f"Вибіркове стандартне відхилення: {s:.2f}")
print(f"Кількість спостережень: {n}")

Вибіркове середнє: 132.29
Вибіркове стандартне відхилення: 10.98
Кількість спостережень: 7


In [3]:
res = stats.ttest_1samp(time_to_purchase, popmean=mu0, alternative='greater')

alpha = 0.10
df = len(time_to_purchase) - 1
t_crit = stats.t.ppf(1 - alpha, df)

print(f"t = {res.statistic:.3f}")
print(f"p-value = {res.pvalue:.5f}")
print(f"t_crit = {t_crit:.3f}")

t = 2.069
p-value = 0.04202
t_crit = 1.440


In [4]:
if res.pvalue < alpha:
    print(f"Відхиляємо H0: середній час до покупки користувачів збільшився на рівні значущості {alpha}")
else:
    print("Не відхиляємо H0: доказів недостатньо")

Відхиляємо H0: середній час до покупки користувачів збільшився на рівні значущості 0.1


На рівні значущості $\alpha = 0.10$ є статистично значущі підстави вважати, що після редизайну сайту середній час до покупки збільшився.
Отже, редизайн, ймовірно, подовжив час до покупки користувачів.

In [7]:
alpha_2 = 0.05

if res.pvalue < alpha_2:
    print("При alpha = 0.05: відхиляємо H0, середній час до покупки користувачів збільшився.")
else:
    print("При alpha = 0.05: не відхиляємо H0, доказів недостатньо.")

При alpha = 0.05: відхиляємо H0, середній час до покупки користувачів збільшився.


Рішення не зміниться при зміні рівня значущості на $\alpha = 0.05$, оскільки p-value (0.042) залишається меншим за 0.05. Отже, нульова гіпотеза відхиляється і є статистично значущі підстави вважати, що середній час до покупки збільшився.

**Завдання 2.**

До спеціальної рекламної кампанії **23%** дорослих упізнавали логотип компанії. Після завершення кампанії відділ маркетингу провів опитування: з **1200** випадково відібраних дорослих **311** упізнали логотип.

Перевірте на рівні значущості **3%** ($\alpha=0.03$), чи дають ці дані достатні підстави стверджувати, що **тепер більше ніж 23%** дорослих упізнають логотип компанії. Для розвʼязку використовуйте бібліотеку `statsmodels`.

Зробіть висновок, чи зросла впізнаваність логотипу.

Додатково, обчисліть довірчий інтревал на заданому рівні значущості і проінтерпретуйте текстом - як він додатково пояснює прийняте нами рішення?

In [10]:
import statsmodels.api as sm
from statsmodels.stats.proportion import proportions_ztest, proportion_confint

count = 311
nobs = 1200
p0 = 0.23
alpha = 0.03

stat, pval = proportions_ztest(count, nobs, p0, alternative='larger')
confint = proportion_confint(count, nobs, alpha=0.03, method='normal')

print("🔹 Тест для однієї пропорції")
print(f"Z-статистика = {stat:.3f}")
print(f"p-value = {pval:.4f}")
print(f"97% довірчий інтервал: {confint}")

if pval < alpha:
    print("Відхиляємо H0: є статистично значущі підстави вважати, що впізнаваність зросла")
else:
    print("Не відхиляємо H0: недостатньо доказів")

🔹 Тест для однієї пропорції
Z-статистика = 2.306
p-value = 0.0106
97% довірчий інтервал: (0.23171700302179205, 0.28661633031154127)
Відхиляємо H0: є статистично значущі підстави вважати, що впізнаваність зросла


Односторонній тест

   * $H_0: p = 0.23$
   * $H_a: p > 0.23$
   * Оскільки $p\text{-value} = 0.0106 < 0.03$, відхиляємо $H_0$.
     Є статистично значущі підстави вважати, що впізнаваність логотипу зросла і тепер **більша за 23%**.

97% довірчий інтервал (0.231, 0.286)

   * Це означає, що з імовірністю 97% справжня частка лежить у межах **від 23.1% до 28.6%**.
   * Оскільки значення 0.23 не входить до цього інтервалу, це також свідчить про статистично значуще зростання впізнаваності.
   * Таким чином, результати довірчого інтервалу підтверджують висновок тесту.